### ISSUE TITLE
spurious /user/* -> /hub/* redirects from jupyterlab browser session when user server stopped

### ISSUE THREAD

**Issue by clkao** [jupyterhub/jupyterhub#2294]
When the user server is stopped, all requests to chp on /user/* that are supposed to hit user server are redirected to the hub, and then redirected to /hub/user/*, this includes:

- kernel sessions: /user/*/api/sessions
- terminal sessions: /user/*/api/terminals
- metrics (if nbresuse and jupyterlab-statusbar is enabled): /user/*/metrics

The first two are actually short-circuited by #2001 so they are just 404 and flood the logs, but the 3rd one gets redirected to the user's spawner page html, causing performance penalties if you are using auth_state to render spawn option menu, as described in https://github.com/yuvipanda/nbresuse/issues/11

This behaviour seems to cause other issues with jupyterlab left open, such as https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/918, and the obscure messages saving notebooks (the PUT /user/*/api/contents request gets a is a 405 response with html content):
<img width="608" alt="2018-11-09 1 19 33" src="https://user-images.githubusercontent.com/69736/48215652-ac123a80-e3bd-11e8-9dcd-a1cb99211f70.png">

Does it make sense to add a default /user/* route in chp to 404 or 503 for non-running users? except the ones that might really be in the browser (such as /user/*/lab, /user/*/tree) that should really be redirected to hub if user not running as they currently do.

**Comment by clkao** [comment]
interim fixes for z2jh-based install with nbresuse:

```
  extraConfig:
    metrics-route-fixes: |
      import json
      from jupyterhub import handlers
      from jupyterhub.handlers import BaseHandler

      class FixupMetricsHandler(BaseHandler):
        async def get(self, user_name):
          self.set_status(404)
          self.set_header("Content-Type", "application/json")
          self.write(json.dumps({"message": "%s is not running" % user_name}))
          self.finish()

      handlers.default_handlers[:0] = [
        (r"/user/(?P<user_name>[^/]+)/metrics", FixupMetricsHandler),
      ]

```

**Comment by minrk** [comment]
FWIW, none of these redirects are spurious. This is the intended behavior for a request to a not-running server, which is meant to start that server back up again. When users ask for servers that aren't running, they should be started.

> Does it make sense to add a default /user/* route in chp to 404 or 503 for non-running users? 

Definitely not. It is super important that users visiting `/user/name/...` trigger spawn and launch their server.

There is an implementation question about *how* that spawn is triggered, and I am inclined for it to be an API request from javascript, so that the launch does not occur automatically for XHRs from left-open JupyterLab. That should get us the best result, where users who visit a URL for a not-running server see "your server is starting up..." but API requests don't trigger the launch.

**Comment by clkao** [comment]
agreed. let me rephrase what I was thinking:

1. The known user urls that might be from browser (/user/*/lab, /user/*/tree, and other entry point) should behave exactly the same as they do.

2. but other requests, which can be determined either by url or Accept header that are known to be XHR, should just stop at chp.

**Comment by minrk** [comment]
CHP can't resolve things that fine-grained, though. And I think it should be the opposite - known API requests should fail with a "your server's not running", but all others should be assumed to be real page requests and result in a spawn.

**Comment by minrk** [comment]
Looks like jquery and friends set x-requested-with, but this is merely a weak convention, and not set by XHR itself. Since jupyterlab uses the lower-level XHR API, it doesn't set this header. Maybe it should, though.

**Comment by manics** [comment]
I'm closing this as it's quite old and there's no concrete action point. If you're still having problems please post on the [Jupyter Community Forum](https://discourse.jupyter.org/) or open a new issue. Thanks!
